Blog post: histograms vs summaries (#2660)
open-telemetry/opentelemetry.io: content/en/blog/2023/histograms-vs-summaries/index.md

cardinality. By comparison, the summary is only a single timeseries for the
precomputed `p99` value. Second, they have very low and configurable relative
error rates. In the histogram example above, we had a potential relative error
of 17% where our summary is guaranteed to be within ± 0.5% accuracy.

# So which should you choose?

The disappointing answer is "it depends," and there is no one-size-fits-all
solution. If you need to aggregate data from many sources, then histograms may
be the right choice. If you are collecting a large number of separate metrics
with very strict SLOs, or your Prometheus server is particularly resource
constrained, then maybe summaries are the right choice for you. Maybe your ideal
solution is a hybrid with some histograms for flexible querying and some
summaries for high-accuracy, low-cost alerting. Only you can know the ins and
outs of your own system and design an observability solution around it that is
accurate and flexible and fits your particular needs. The key is knowing the
strengths and limitations of the available tools so you can make informed
decisions.

# Bonus round: native histograms

> I'm planning a longer post on this so I'll keep this short, but many of the key
> disadvantages of histograms are mitigated by native histograms, called
> exponential histograms in OpenTelemetry. Available in Prometheus as an
> experimental feature since v2.40.0, and stable in the OpenTelemetry
> specification as of v1.17.0, native histograms enable very efficient data
> collection and transmission, fewer, and a constant number of, timeseries created
> per histogram, and very low relative error rates. They achieve these and other
> benefits by defining buckets automatically according to a scale factor and
> resizing intelligently as needed. If you're not happy with the state of your
> current histograms and summaries, I encourage you to give native histograms a
> try. As of this writing there are no official Prometheus docs on native
> histograms, but if you stay tuned I plan to add a thorough explanation of them
> in the coming days.

Reviewer: ```suggestion
I'm planning a longer post on this so I'll keep this short, but many of the key
disadvantages of histograms are mitigated by native histograms, called
exponential histograms in OpenTelemetry. They are available in Prometheus as an
experimental feature since v2.40.0, and stable in the OpenTelemetry
specification as of v1.17.0.  Native histograms come with several advantages:

 * Very efficient data collection and transmission
 * A constant number of timeseries created (and fewer of them) per histogram
* Very low relative error rates.
* Automatic bucket boundaries, making ergonomics of using them easier

These advantages are accomplished because they automatically define bucket
boundaries according to a scale factor and intelligent resizing.
If you're not happy with the state of your current histograms and summaries,
I encourage you to give native histograms a try. As of this writing there are no
official Prometheus docs on native histograms, but if you stay tuned I plan to
add a thorough explanation of them in the coming days.
```

Reviewer: I like it!

---
title: Histograms vs Summaries
date: 2023-05-15
author: '[Daniel Dyla](https://github.com/dyladan)'
spelling: cSpell:ignore Dyla quantile quantiles timeseries aggregatable
canonical_url: https://dyladan.me/histograms/2023/05/03/histograms-vs-summaries/
---

In many ways, histograms and summaries appear quite similar. They both roll up
many data points into a data structure for efficient processing, transmission,
and storage. They can also both be used to track arbitrary quantiles such as the
median or p99 of your data. So how do they differ? Let's dive in.

> # Histograms

Reviewer 2: In-text headings start at level 2:

```suggestion
## Histograms
```

bucket (`300 <= x < 500`) by summing the bucket counts until we exceed that
rank. Finally, using relative rank within the bucket of 24 (`1134 - 1110`), we
can estimate the p90 value to be 360ms (`300 + ((24 / 80) * (500 - 300))`) using
linear interpolation. It is important to know that this is an _estimation_ and
could be off by as much as 60ms (`360 - 300`), a relative error of 17%
(`60 / 360`). This error can be mitigated by configuring more and smaller
buckets around your SLO values, but never eliminated.

One important property of histograms is that they are _aggregatable_, meaning
that as long as the bucket boundaries line up, an arbitrary number of histograms
can be combined into a single histogram with no loss of data or precision. This
means that an arbitrary number of hosts can report histogram data structures to
a server, which can aggregate and compute quantiles from all of them as if they
were reported by a single host. By collecting histograms from 1 or more hosts
over a long period of time, developers can gain a strong understanding of how
their data is distributed and how that distribution changes over time.

> # Summaries

Reviewer 2: ```suggestion
## Summaries
```

is more complex so it won't be discussed here, but one important distinction is
that the value is calculated on the client before it is sent to the server. The
most important consequence of this is that summaries from multiple clients
_cannot be aggregated_. Another disadvantage is that if you cannot query
arbitrary φ-quantiles, only those which you have configured and collected in
advance.

Given these disadvantages, summaries do have some advantages. First, they trade
off a small performance penalty on the client for a significant reduction in
transmission, storage, and server processing cost. In our histogram example
above, the distribution is represented as 12 separate timeseries: 1 counter for
each bucket + 1 bucket for out of range values + a total sum of all values. That
is for a single, relatively modest, histogram with no attributes to multiply
cardinality. By comparison, the summary is only a single timeseries for the
precomputed `p99` value. Second, they have very low and configurable relative
error rates. In the histogram example above, we had a potential relative error
of 17% where our summary is guaranteed to be within ± 0.5% accuracy.

> # So which should you choose?

Reviewer 2: ```suggestion
## So which should you choose?
```

The disappointing answer is "it depends," and there is no one-size-fits-all
solution. If you need to aggregate data from many sources, then histograms may
be the right choice. If you are collecting a large number of separate metrics
with very strict SLOs, or your Prometheus server is particularly resource
constrained, then maybe summaries are the right choice for you. Maybe your ideal
solution is a hybrid with some histograms for flexible querying and some
summaries for high-accuracy, low-cost alerting. Only you can know the ins and
outs of your own system and design an observability solution around it that is
accurate and flexible and fits your particular needs. The key is knowing the
strengths and limitations of the available tools so you can make informed
decisions.

> # Bonus round: native/exponential histograms

Reviewer 2: ```suggestion
## Bonus round: native/exponential histograms
```
